php/libraries/Database.class.inc

```diff
@@ -796,6 +869,82 @@ class Database
+     *
+     * @throws LorisException     If the supplied key is empty or null
+     * @throws DatabaseException  If the key is not part of the query itself or
+     *                            if there are not exactly 2 columns selected
+     * @throws DatabaseException  If the key is not unique within the resulting set
+     *
+     * @return array Associative array of form uniqueKey=>value containing all
+     *               value for the non-uniqueKey element of the select statement
+     */
+    function pselectColWithIndexKey($query, $params, $uniqueKey)
+    {
+        if (is_null($uniqueKey) || empty($uniqueKey)) {
+            throw new LorisException(
+                "The pselectColWithIndexKey() function expects the uniqueKey
+                 parameter to not be null or empty. If re-indexing on the primary 
+                 key is not necessary please use the pselectCol() function instead."
+            );
+        }
+
+        $result = $this->pselect($query, $params);
+
+        if (empty($result)) {
+            return $result;
+        }
+
+        $filteredResult = array();
+        // re-order the return array
+        foreach ($result as $row) {
+            $colNumber = count($row);
+            // Check first that the row contains the primary key supplied
+            if (!array_key_exists($uniqueKey, $row) || $colNumber !== 2) {
+                throw new DatabaseException(
+                    "The query supplied to pselectColWithIndexKey() should only 
+                    contain the unique key and one other column in the SELECT 
+                    clause. Make sure to supply the appropriate key in the SELECT 
+                    statement to match the supplied parameter of this function.",
+                    $query
+                );
+            }
+
+            // Check that the primary key is indeed unique to avoid overriding data
+            // in the result array
+            if (isset($filteredResult[$row[$uniqueKey]])) {
+                throw new DatabaseException(
+                    "The uniqueKey supplied to pselectColWithIndexKey() does not 
+                    appear to be unique or is nullable. This function expects the 
+                    key to be both UNIQUE and NOT NULL.",
+                    $query
+                );
+            }
+
+            // Store the value of the key for this specific row then unset that
+            // value from the $row array so that the only element remaining is
+            // the column desired and its value for this row
+            $uniqueKeyValue = $row[$uniqueKey];
+            unset($row[$uniqueKey]);
+
+            // Note: reset() rewinds array's internal pointer to the first element
+            // and returns the value of the first array element, in this case the
+            // desired column value
+            $filteredResult[$uniqueKeyValue] = reset($row);
+        }
+
+        return $filteredResult;
+    }
+
```

Comment:
Now that I think of it, you're right. Doesn't pselectCol always return a single column? Isn't this just requiring an extra parameter @ridz1208?

Reply:
The idea here...
Imagine the following table (notice that the primary key is an autoincrement but they are not consecutive to distinguish them from row numbers)
![screenshot from 2018-02-05 10-51-16](https://user-images.githubusercontent.com/15268287/35815127-cb23986a-0a65-11e8-9112-706f3e4a826e.png)

Ideally what you want and what is used all throughout the loris codebase is an array of the form `$centerID=>$centerName` as follows

```
$DB->pselectColWithIndexKey("SELECT CenterID, Name FROM psc WHERE CenterID<11",array(),"CenterID")
Array
(
    [1] => CCNA Testing Site
    [5] => True North Clinical Research (Halifax)
    [6] => St. Joseph's Hospital
    [7] => Centre for Diagnosis and Research on Alzheimer's Disease (CEDRA), Memory Clinic
    [8] => Centre Hospitalier de l'Université de Montréal
    [10] => Douglas Mental Health University Institute, Memory Clinic
)
``` 

But currently we can only get one of the following... in either case we will need a secondary forloop to extract the key=>value association we are looking for.

```
$DB->pselect("SELECT CenterID, Name FROM psc WHERE CenterID<11",array())
Array
(
    [0] => Array
        (
            [CenterID] => 1
            [Name] => CCNA Testing Site
        )

    [1] => Array
        (
            [CenterID] => 5
            [Name] => True North Clinical Research (Halifax)
        )

    [2] => Array
        (
            [CenterID] => 6
            [Name] => St. Joseph's Hospital
        )

    [3] => Array
        (
            [CenterID] => 7
            [Name] => Centre for Diagnosis and Research on Alzheimer's Disease (CEDRA), Memory Clinic
        )

    [4] => Array
        (
            [CenterID] => 8
            [Name] => Centre Hospitalier de l'Université de Montréal
        )

    [5] => Array
        (
            [CenterID] => 10
            [Name] => Douglas Mental Health University Institute, Memory Clinic
        )

)
```
```
$DB->pselectCol("SELECT Name FROM psc WHERE CenterID<11",array())
Array
(
    [0] => CCNA Testing Site
    [1] => True North Clinical Research (Halifax)
    [2] => St. Joseph's Hospital
    [3] => Centre for Diagnosis and Research on Alzheimer's Disease (CEDRA), Memory Clinic
    [4] => Centre Hospitalier de l'Université de Montréal
    [5] => Douglas Mental Health University Institute, Memory Clinic
)
```
```
$DB->pselectWithIndexKey("SELECT CenterID, Name FROM psc WHERE CenterID<11",array(),"CenterID")
Array
(
    [1] => Array
        (
            [Name] => CCNA Testing Site
        )

    [5] => Array
        (
            [Name] => True North Clinical Research (Halifax)
        )

    [6] => Array
        (
            [Name] => St. Joseph's Hospital
        )

    [7] => Array
        (
            [Name] => Centre for Diagnosis and Research on Alzheimer's Disease (CEDRA), Memory Clinic
        )

    [8] => Array
        (
            [Name] => Centre Hospitalier de l'Université de Montréal
        )

    [10] => Array
        (
            [Name] => Douglas Mental Health University Institute, Memory Clinic
        )

)
```

I think this function might just be a better version of `pselectCol` but I dont think it is redundant with `pselectWithIndexKey` because it returns a flatened array directly useable by the code requesting it

Reply:
@driusan did the explanation make it clearer ?

Reply:
Other than the fact that I didn't see the comment until now, yes.